### DISCUSSION TITLE
Ender3 V2/S1 Professional Firmware 20220527

### DISCUSSION THREAD

**Discussion by mriscoc** [mriscoc/Ender3V2S1#172] [Announcements]
Universal 20220527 Maintenance release
==========================

## What's new in this release ##
- Improved mesh tilting algorithm.
- A new custom G-code C104 U1 T### to unlock the hotend maximum temperature to ###°C (CAUTION) 
- Fix a little cosmetic bug in Advanced Pause (change filament)
- Some status messages clear by themselves after 30 s
- Increase small movements (0.01mm) accuracy
- Increase G30 Z precision
- Enhanced Edit Mesh menu
- Enhanced redraw menu while an item is being edited
- Add M503 support for Custom G-codes
- Better calculus of probing area for the bed tramming wizard
- Enhanced Smart Fill Mesh
- Validation of the UBL mesh when it is loaded (through menu or G-code)
- Debugging thumbnail data validation messages
- New BLTouch reset item in Probe settings menu
- Automatic mesh inset and park head position corrections after load settings and change physical parameters
- Official support for Ender-3 S1F4 and Ender-3 V2 SKRMINI-E3V3
- Merged the latest Marlin bugfix

 This firmware is based on Marlin, has been optimized and has enhancements for the Creality Ender 3 V2/S1 printer. This is an **universal firmware** that can be used in both RET6 and RCT6 SoC chips that have 256k or 512k Flash memory. 

## Versions
- **Ender 3v2 MM**: if you have the Ender 3v2 near to stock then use this version (Manual Mesh).
- **Ender 3v2 BLT**: use this version if you installed a CR/3D/BLTouch level sensor in the Ender 3v2.
- **Ender 3S1**: use this version if you have an Ender 3S1 printer

### CAUTION! It is confirmed that Creality is selling some S1 with a STM32F4 chips, The current firmware version is incompatible with this SoC   
Only S1 boards with **STM32F1xx** are supported. Currently the S1 STM32F4 support is experimental and a preliminar precompiled S1 F4 binary can be found here: https://github.com/mriscoc/Ender3V2S1/issues/128

## UBL versions
Binaries with UBL in their filename have support for [Marlin Unified Bed Leveling](https://marlinfw.org/docs/features/unified_bed_leveling.html), this system is similar to the bilinear leveling but with a few extra features to provide users with more control. There are multiple guides on the web about UBL, we recommend reading about it before using this new leveling system.

* https://marlinfw.org/docs/features/unified_bed_leveling.html
* https://all3dp.com/2/marlin-ubl-unified-bed-levelin-simply-explained/
* https://www.3dmakerengineering.com/blogs/3d-printing/unified-bed-leveling-marlin
* https://hackaday.com/2022/01/14/3d-printering-getting-started-with-universal-bed-leveling/
  
<img align="left" src="https://raw.githubusercontent.com/mriscoc/Ender3V2S1/Ender3V2S1-Released/screenshots/main.jpg" width="270" height="480" />
<img  src="https://raw.githubusercontent.com/mriscoc/Ender3V2S1/Ender3V2S1-Released/screenshots/main-iconset.jpg" width="270" height="480" />

The binary files of this release are up to date and you can download them from the assets at the bottom of this page. To know if you have the last release go to the info page in the display of your printer and compare the release build date with the name of the binary file. The filename follows the convention:  
  
Ender3v2-422MM-**YYYYMMDD**.bin
  
Where YYYYMMDD indicates the date of the build. 422 refers to the mainboard version **4.2.2**, 423 refers to the mainboard version **4.2.3**, and 427 is for the **4.2.7** mainboard version, if you got a **4.3.1** board use the 422 version. 

![Identificando_version_de_placa](https://user-images.githubusercontent.com/2745567/168324748-0c6dca47-41f3-4ad5-9146-a1a3caf6a7ce.jpg)

## Special versions
There are special compiles (as V6/Volcano) here: https://github.com/mriscoc/Marlin_Configurations/releases

## Community links
* [Telegram](https://t.me/ender3v2s1firmware)
* [Reddit](https://www.reddit.com/r/Ender3v2Firmware) 
* [Ender 3V2 Facebook](https://www.facebook.com/groups/ender3v2firmware)
* [Ender 3S1 Facebook](https://www.facebook.com/groups/ender3s1printer)

## Donations
Thank you for your support, I receive donations through [Patreon](https://www.patreon.com/mriscoc) and [Paypal](https://www.paypal.com/paypalme/mriscoc)   

[<img src="https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif">](https://www.paypal.com/donate?business=85SPAAR6UZEE8&currency_code=USD)

# Firmware features

* Universal RCT6/RET6 Ender 3 V2/S1 Support
* DWIN / DACAI screens support
* Improved support for Manual Mesh and CR/3D/BLTouch
* User configurable power lost recovery enabler
* Manual and assisted bed tramming
* Tramming Wizard for BL/3D/CRTouch versions
* Long filename support with scroll
* Baby Z-Stepping
* Firmware Retract
* Advanced pause feature
* Configurable park-head position (default x=240 y=220)
* Filament change support
* Filament management (load/unload/park head) in prepare menu
* Print Flow on Motion and Tune menu 
* Home axis after motor disabling
* Fastest Z feed rate
* Live move for all motors but extruder
* Support for S-curve Acceleration
* Reboot printer item on control menu
* Increase extruder feed rate limit to 150 for BMG users
* Allows to invert direction of the extruder on Motion menu
* Support for BTT Smart filament sensor
* Selectable runout filament sensor active state (LOW/HIGH/MOTION)
* Editable runout distance
* Support for µSD extensions
* PID popup with temperature plot
* Lock screen
* Editable physical dimensions
* Mesh viewer
* Custom pre-heat
* Enable/disable beeps sound
* New print g-code preview (use [this ](https://github.com/mriscoc/Ender3V2S1/wiki/How-to-generate-a-gcode-preview)guide)
* New print done screen
* Two decimal positions for axis coordinates value
* New Z-Offset wizard
* New PID settings menus
* PID moves the head to the center for more accuracy
* Leveling bed temperature linked to bed PID temperature
* Editable extrude minimum temperature
* Brightness settings on tune menu
* Turn off the backlight of the lcd in the tune menu
* New status line system, allows the scroll of large messages
* Home submenu with individual axis homing
* Enabled user setting for 115200 bauds (Creality Box compatibility)
* Head parking before recovery after power failure
* Offset icon changes to show leveling status
* Live end-stop diagnostic page
* Move bed to front after cancelling a print.
* Editable firmware retracts settings for Tune and filament settings menu
* Print Statistics page accessible from the Advanced Settings menu
* Toolbar with 11 configurable shortcuts actions.
* Support in source code of control menus for LED lights as the Neopixel

![latest release](https://raw.githubusercontent.com/mriscoc/Ender3V2S1/Ender3V2S1-Released/screenshots/New-beta.gif)

### Manual mesh
* Assisted manual bed tramming
* Assisted manual mesh leveling menu
* Selectable mesh points from 2x2 to 9x9
* Mesh viewer
* Persistent Z-offset
* Move Z to 0 and disable motors option
* Mesh editor for fine tuning

### CR/3D/BLTouch
* Safe Probe Z-Homing
* Bed auto level
* High speed probe
* Retry bed level on error
* Mesh viewer
* Assisted tramming
* Selectable mesh points from 2x2 to 9x9
* Allows the user to cancel an auto leveling process 
* Increase Max Probe Offset and Probing Margin to 60 mm
* M48 Probe Test to Probe Settings menu
* Probe Stow/Deploy in Probe Settings menu
* Option to disable HS mode for better compatibility with CRTouch
* Option to change the Z Probe feed rate
* Configurable Bed auto leveling temperature
* Mesh editor for fine tuning
* Automatic maximize leveling area
* Automatic center leveling area

>Use the 2x2 mesh to level you bed!

<a href="https://youtu.be/qzYJC2AHbW8"><img src="https://img.youtube.com/vi/qzYJC2AHbW8/maxresdefault.jpg" width="640" height="400" /></a>  
  
### G-Codes enhanced support
* Filament Load Unload G-codes (M701/M702)
* Support for M600 filament change
* Support for G-code M117 messages
* Enable M575 change baud rate by G-code
* Extended M75 with filename support, use M75 {{ event.name }} in g-code script
* Enable ARC and 'P' parameter to specify complete circles (G2/G3)
* G35 support
* M155 Geometry report
* Native support of marlin brightness and M256 g-code

### Support for special nonstandard custom C G-Codes
* Implemented C11 En Rn Gn Bn gcode for colorize UI elements
* C108 to cancel screens waiting for user simple confirmation
* C29 N# to set amount of grid points of the mesh leveling
* C100 X### Y### to setup minimum limits
* C101 X### Y### z### to setup maximum limits
* C102 X### Y### to setup bed size
* C125 X### Y### z### to setup the park position
* C412 M# set run-out sensor active state (M0:LOW, M1:HIGH, M2:MOTION)
* C510 to lock/unlock the screen (C510 U1 to unlock)
* C562 E# to invert the Extruder (E1 to invert)
* C851 M### F#### to set probing margin, Z feed rate

<img src="https://raw.githubusercontent.com/mriscoc/Ender3V2S1/Ender3V2S1-Released/themes/Select_Colors.jpg" width="270" height="480" /> 

### Advanced settings menu
* Home XYZ offsets menu
* Probe offsets menu (BLTouch version)
* Customizable park-head position
* Customizable bed size and z heigh up to 999 x 999 x 999
* Launch PID for nozzle and bed with graphical feedback
* Filament run-out sensor with active state setup (LOW/HIGH)
* Filament load and unload lengths menu items in run-out menu
* Power-loss recovery enable/disable option
* LCD Brightness control
* Select Color Sub Menu
* Beep/Sound enable/disable option
* Selectable mesh point (BLTouch version)
* Mesh viewer
* Lock screen

### Special Octoprint / Host print support
* High speed (250000 baud) serial communication
* In screen control of host printing (M73, M75-M77) 
* Display messages on status line (M117)
* Color change of UI elements for visual status report (C11)
* Binary file transfer for remote firmware upgrade
* List binary (.bin) files (M20) for flash firmware support with BFT
* Arc Welder Octoprint plugin support
* EEPROM Editor Octoprint plugin support
* Host action commands for filament runout
* Emergency Parser active
* Advanced OK

**Octoprint bed visualizer support**
<img src="https://raw.githubusercontent.com/mriscoc/Ender3V2S1/Ender3V2S1-Released/screenshots/Bed-visualizer.png" width="350" height="225" />

### Display enhancements
* Extended DWIN library
* Enhanced status messages
* Improved status bar with temperature, flow, fan, z-offset and xyz positions
* PID auto tune progress on display with plot
* Leveling progress on display
* Boot screen for M997
* Transparent icons for better color theme adaptability
* New print g-code preview
* New print done screen

**A lot of additions and fixes from Bugfix development branch**

## More info on the wiki pages ##
- [**How to install the firmware**](https://github.com/mriscoc/Ender3V2S1/wiki/How-to-install-the-firmware)

- [**Octoprint settings**](https://github.com/mriscoc/Ender3V2S1/wiki/Octoprint)

- [**How to make the Manual Mesh**](https://github.com/mriscoc/Ender3V2S1/wiki/Manual-Mesh)

- [**How to install a 3D/BLTouch**](https://github.com/mriscoc/Ender3V2S1/wiki/3D-BLTouch)

- [**How to apply a color theme**](https://github.com/mriscoc/Ender3V2S1/wiki/Color-Themes)

https://user-images.githubusercontent.com/2745567/143180943-f69f5967-34c1-421f-86a1-466ecccab97d.mp4

## Precompiled firmware

<hr /><em>This discussion was created from the release <a href='https://github.com/mriscoc/Ender3V2S1/releases/tag/20220527'>Ender3 V2/S1 Professional Firmware 20220527</a>.</em>

**Comment by anilgunduzz** [comment]
Hi,

Can you add also SKR Mini E3 V3 config files to the Configurations folder? I would like to try to activate linear advance.

**Reply by mriscoc** [reply]
> Hi,
> 
> Can you add also SKR Mini E3 V3 config files to the Configurations folder? I would like to try to activate linear advance.

Check here: https://github.com/mriscoc/Special_Configurations

**Reply by anilgunduzz** [reply]
Ah, sorry. Thank you.

**Comment by rubberbandboy** [comment]
Hi,

Do we get an S1 release without UBL? (f1)

**Reply by mriscoc** [reply]
Only S1 F1 can have UBL support, the F4 board doesn't have enough flash space to support UBL without disabling other features.  I just posted a F1 ABL version too.

**Comment by SjoukeAlthuis** [comment]
Thank you for this great firmware. With this firmware printing and leveling is made easy.

**Comment by JRC3052** [comment]
Yes, Thank you for this great firmware. This version no longer has the "Set Home Offsets" option. Anyway you could add that back on the next version? Thanks

**Reply by mriscoc** [reply]
Please, check this: https://github.com/mriscoc/Ender3V2S1/wiki/How-to-set-the-home-offsets

**Comment by kiu345** [comment]
Something went wrong with the source archives, they contain the github page, not the source code.
![image](https://user-images.githubusercontent.com/4389276/172058901-2604e8de-59e9-4d06-abae-177d9b9c9bcd.png)

**Reply by mriscoc** [reply]
Download from the repository: https://github.com/mriscoc/Ender3V2S1

**Comment by bee-mann** [comment]
Great work @mriscoc 
Do you perhaps have firmware for Ender 3v2 427 mainboards (512 editions) with bltouch (V3.1) and and Sprite all Metal extruder (300 degrees C)?
I am currently using 427 Jyers firmware but max temps are 275.

Thanks in advance for your reply.

**Reply by mriscoc** [reply]
For 4.2.7: [Ender3V2-427BLT-UBL-AM3950-20220420.bin](https://github.com/mriscoc/Special_Configurations/releases/download/AM3950T13/Ender3V2-427BLT-UBL-AM3950-20220420.bin) 
Max temp is 300 - 15 = 285°C, read here: https://github.com/mriscoc/Special_Configurations/releases/tag/AM3950T13

**Reply by bee-mann** [reply]
Thank you so much for this!

**Reply by bee-mann** [reply]
Hi

Thank you, been using your firmware for a month and it's absolutely fantastic.
I have an enhancement request I have Z steppers and want to find out if you can add this feature to help me auto-align my Z.
**G34 - Z Steppers Auto-Alignment**
https://marlinfw.org/docs/gcode/G034-zsaa.html

Thank you

**Comment by PapsDizzy** [comment]
will this firmware work with gd32f303ret6 chip on 4.2.2 board?

**Reply by mriscoc** [reply]
This SoC is not officially supported by Marlin but could work with it.

**Reply by PapsDizzy** [reply]
ok so try and hope that no erros are coming

**Comment by divide-mike** [comment]
How can I change the language to Chinese？

**Reply by mriscoc** [reply]
Currently only CrealityUI and ProUI on official Marlin support Chinese.

**Comment by divide-mike** [comment]
How to change the temperature to 300 degrees

**Reply by mriscoc** [reply]
You must change the thermistor to one that has a thermistor table for higher temperatures. Currently, the stock thermistor only supports up to 285 (300-15) °C.

**Comment by divide-mike** [comment]
Does the g code need to be compiled and re-burned in the marlin firmware?

**Reply by mriscoc** [reply]
No, special G-codes work right with the current precompiled versions. But you need to use a terminal with support for non-standard G-codes. https://github.com/mriscoc/Ender3V2S1/wiki/Custom-C-gcodes

**Comment by Lucastodu** [comment]
Hi, a friend of my have a Ender 3 Max, so he wants to know if it's possible to use this firmware on it? 

Just to clarify things here , his printer has the same board as the one in the Ender 3 V2, more specific the 4.2.2 board.

Can anyone give any coments?

Thank you in advance.

**Reply by Nbrooks639** [reply]
i believe this would work just fine with that board.  From the lcd you will have to adjust all the axis differences but its pretty simple.

**Reply by Lucastodu** [reply]
Thank you for the answer!

I forgot to mention that his display is the simple one like the original ender 3, probably it will need this to be changed in the configuration file right?!

**Comment by Creat-new** [comment]
Hello, I am very grateful see your work, but I hope you can help us to modify the ENDER 3 S1 PRO, because S1 does not seem to work。 @mriscoc

**Reply by mriscoc** [reply]
No support for touchscreens

**Reply by MichiMess** [reply]
I have the Ender pro and changed the the touchscreen for the V2 screen for this firmware. Much, much better!

Read this for more info:

https://www.reddit.com/r/ender3/comments/u03vg8/firmware_ender_3_s1_pro/ieghpo6

**Reply by Creat-new** [reply]
This is a good idea, but the screen size and resolution of the V2 and S1 are still different and need to be dealt with @MichiMess

**Reply by MichiMess** [reply]
You mean the touchscreen from the S1 pro vs the V2 screen? Yeah, that's a downgrade... But it's so much better with this firmware. I hoping for a firmware so we can use the touchscreen... But for now I'm using the V2 screen.

**Reply by Creat-new** [reply]
i think so, thanks for all of your contributions, it is of course the best to have the firmware to directly upgrade the touch screen

**Comment by EXCISIONMODZ** [comment]
add the led menu pls
i cant get it to compile

**Comment by MichiMess** [comment]
I have the F4 chip. But I can't find UBL for F4.  Is that right?

**Reply by mriscoc** [reply]
Currently there is not a precompiled UBL version for F4 due to lack of program memory.

**Reply by MichiMess** [reply]
Ok, understand. But what a great firmware! Thanks for this!!

**Comment by Beastboy1989** [comment]
Which firmware would I select for the EZable to operate properly?

**Reply by mriscoc** [reply]
Maybe this: https://github.com/mriscoc/Special_Configurations/releases/tag/EZABL

**Reply by Beastboy1989** [reply]
Thank you I must have completely over looked it.

**Comment by Creat-new** [comment]
thanks for much  such a good firmware, I solved the screen incompatibility problem of S1 PRO by changing the V2 screen, but there is still a flaw in that the printing temperature is only 275 and cannot reach 315, how should I change it What about your source code? @mriscoc

**Reply by mriscoc** [reply]
I'm working in the current source after a big mess when merged with latest Marlin Bugfix 2.1 and overwrote my working copy, there is an old copy if you need here: https://github.com/mriscoc/Ender3V2S1/tree/20220419

**Reply by Creat-new** [reply]
thanks

**Comment by TropicalCyclone** [comment]
is there a version of this for the F4 where it has the laser engraving mode with it?

**Reply by mriscoc** [reply]
No, I would need to have the laser engraver module for try to make something compatible with that.

**Reply by TropicalCyclone** [reply]
thats a bummer, that means id have to flash to the default firmware to use the laser cutter

**Comment by jerson205** [comment]
Is it possible to get this firmware with 300 degrees for the Ender 3s1 pro ? Or How to  customize it to have it? I’ve got an s1 pro. Thanks a lot, very good work

**Reply by MichiMess** [reply]
Your have to change your screen.

https://www.reddit.com/r/ender3/comments/u03vg8/firmware_ender_3_s1_pro/ieghpo6

**Reply by jerson205** [reply]
Its done Yes to version lcd of ender 3v2

**Comment by zombiejesus123** [comment]
is there a filament run out built into the firmware because i cant seem to find the setting for it

**Reply by mriscoc** [reply]
Advanced / Filament Settings

**Reply by zombiejesus123** [reply]
> Advanced / Filament Settings

found it thanks

**Comment by Beastboy1989** [comment]
![Untitled](https://user-images.githubusercontent.com/107695480/178958600-19056501-2502-4afc-a2a4-a75bd560c385.png)
Could someone tell me how to fix this

**Reply by mriscoc** [reply]
New sources will be posted at the weekend

**Comment by GT-Geo** [comment]
That's great news, I'm hoping you were able to address this;
[https://github.com/mriscoc/Ender3V2S1/discussions/214]

**Comment by MNathan19951** [comment]
My build is Ender 3V2 (4.2.2) + CR Touch + Sprite Pro Extruder and I am confused about which Professional Firmware is compatible with my setup. Z-offset has been properly set after each firmware load (Auto Home, Z-offset, bed level).

I originally loaded the official creality version for sprite extruder and worked just fine [Ender-3 V2-HW-V4.2.2-SW-V1.0.5T-SpritExt-20220714.bin]
Then, changed to professional firmware and had print issues [Ender3V2-422-BLTUBL-T13-MPC-20221222.bin]
 -Filament not sticking to bed. It would extrude (appeared to be slower?) and curl up towards the nozzle.
 -X/Y offsets had to be adjusted to compensate for the new extruder position.
 -Cleaned glass bed, applied glue stick to bed, changed glass bed, glue stick on that bed, changed filament to fresh spool and leveled with auto bed mesh & tramming wizard in between every step, etc).

Finally gave up and changed back to original creality firmware for sprite on ender 3v2 and print was sticking fine to bed. My theory is the professional firmware assumes standard extruder/nozzle/etc and behaves differently with sprite.

Any help on compiling a professional firmware that works with my build would be appreciated. Or should I just load Ender 3S1 firmware and try that?

**Reply by mriscoc** [reply]
For the Sprite extruder pro, you need to follow this guide: https://github.com/mriscoc/Ender3V2S1/wiki/Calibration-Guides
The T13 in the firmware filename is only to note the support for the Marlin thermistor calibration table 13. But the Sprite extruder pro has other offsets for the nozzle and probe than stock V2 printers.

**Reply by MNathan19951** [reply]
Thanks for the suggestion. 
More specifically, my issue was the e-steps not being set properly. I plugged in the default values from the Creality firmware (ender3V2+Sprite) to get me in the right ballpark.

**Comment by YashaMars** [comment]
Good afternoon, could you please tell me how to increase the temperature limit from 260 degrees to 300?
I have an Ender 3s1 with a thermistor for 300+ degrees
Firmware installed: PF V:2.1.3 MRiscoC 20230312 - 20:35

**Reply by mriscoc** [reply]
Look for C104

**Reply by YashaMars** [reply]
> Look for C104

Excuse me, please tell me which version I need to download Ender3S1-F4-20230522B.bin or Ender3S1-F4-UBL-20230522B.bin
Honestly I'm not quite sure what UBL is.
And do they have LA support?
If I download the version Ender3S1-F4-LA-20230524.bin, will it support C104 GCod?

**Reply by mriscoc** [reply]
All the information is in the wiki and the links in it.

**Reply by mriscoc** [reply]
> will it support C104 GCod?

Sorry, we recently learned that due to a fixed Marlin bug, current precompiled binaries do not have support for the C104 custom G-code.

**Comment by hasemin** [comment]
Does this software support the laser engraver for the ender 3 s1? Thanks in advance

**Reply by mriscoc** [reply]
Use the CVLM version: https://github.com/mriscoc/Special_Configurations/releases/tag/CVLM
